# Proyek Analisis Data: Bike Sharing Dataset
- **Nama:** Muchammad Wildan Alkautsar
- **Email:** muchammadwr@gmail.com
- **ID Dicoding:** muchammadwr

## Menentukan Pertanyaan Bisnis

**SMART Question** adalah sebuah framework untuk merumuskan pertanyaan secara terstruktur agar memeperoleh informasi yang mendalam.

* **Spesific (Spesifik)**
  * Pertanyaan harus jelas, fokus pada sebuah topik tertentu, dan tidak bermakna ganda. Hindari pertanyaan yang terlalu luas.
    * Salah: Bagaimana cara meningkatkan penjualan?
    * Benar: Faktor apa saja yang memengaruhi penurunan penjualan produk kategori elektronik di wilayah Jakarta selama kuartal terakhir?
* **Measurable (Terukur)**
  * Pertanyaan harus bisa dijawab dengan angka atau matrix yang konkret, harus tahu apa yang akan dihitung.
    * Salah: Apakah pelanggan senang dengan layanan kita?
    * Benar: Berapa skor rata-rata Customer Satisfaction untuk layanan purna jual bulan ini dibandingkan bulan lalu?
* **Action-Oriented (Berorientasi Aksi)**
  * Hasil dari pertanyaan harus bisa memberikan arahan untuk melakukan tindakan nyata. Jika pertanyaan terjawab, stakeholder harus tahu apa langkah selanjutnya.
    * Salah: Mengapa orang suka berbelanja?
    * Benar: Fitur apa pada aplikasi yang paling sering digunakan sebelum pengguna memutuskan untuk melakukan checkout?
* **Relevant (Relevan)**
  * Hasil dari pertanyaan harus sejalan dengan tujuan utama bisnis atau masalah yang sedang dihadapi.
    * Salah: Menanyakan tentang stok gudang saat masalah utamanya adalah efektivitas kampanye media sosial.
    * Benar: Apakah kampanye iklan di Instagram memberikan Return on Ad Spend (ROAS) yang lebih tinggi dibandingkan iklan di TikTok?
* **Time-bound (Terikat Waktu)**
  * Pertanyaan harus ada batasan waktu yang jelas agar analisis memiliki konteks yang tepat.
    * Salah: Berapa banyak pengguna baru kita?
    * Benar: Berapa tingkat pertumbuhan pengguna baru secara bulanan (Month-over-Month) sepanjang tahun 2025?

**Contoh pertanyaan bisnis yang memenuhi seluruh elemen SMART**

***"Faktor apa saja yang memengaruhi penurunan conversion rate pada pengguna aplikasi Android di wilayah Jabodetabek sebesar 5% selama periode Flash Sale Maret 2026?"***

Keterangan:

- **Specific**: Fokus pada "penurunan conversion rate" untuk "aplikasi Android" di "Jabodetabek". Bukan sekadar penjualan turun.
- **Measurable**: Ada angka konkret yang ingin dianalisis, yaitu penurunan sebesar "5%".
- **Action-Oriented**: Dengan mengetahui faktor penyebabnya misalnya bug pada versi Android tertentu atau kendala logistik di Jabodetabek, tim bisa langsung melakukan perbaikan teknis atau operasional.
- **Relevant**: Penurunan konversi saat Flash Sale adalah masalah kritis bagi bisnis retail/e-commerce.
- **Time-bound**: Dibatasi pada periode spesifik "Flash Sale Maret 2026".

- **Question 1:** During the 2012 period, which hours, days, and seasons recorded the highest average bike rentals?

- **Question 2:** How did monthly rental demand change throughout the 2012 period?

- **Question 3:** How did average bike rentals differ across weather conditions during 2012?

- **Question 4:** Which weather-related factor (temperature, humidity, or wind speed) had the strongest relationship with rental demand?

- **Question 5:** How did rental patterns differ between casual and registered users during 2012?

- **Question 6:** Which user segment contributed the largest share of total rentals over the two-year period?


## Import Semua Packages/Library yang Digunakan

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
from datetime import time

warnings.filterwarnings("ignore")

## Data Wrangling

### Gathering Data

#### Load daily_df

In [2]:
daily_df = pd.read_csv("https://raw.githubusercontent.com/muchammadwr/bike-sharing-demand-analysis/refs/heads/main/submission/data/day.csv")
daily_df.head()

,instant,dteday,season,yr,mnth,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
0,1,2011-01-01,1,0,1,0,6,0,2,0.344167,0.363625,0.805833,0.160446,331,654,985
1,2,2011-01-02,1,0,1,0,0,0,2,0.363478,0.353739,0.696087,0.248539,131,670,801
2,3,2011-01-03,1,0,1,0,1,1,1,0.196364,0.189405,0.437273,0.248309,120,1229,1349
3,4,2011-01-04,1,0,1,0,2,1,1,0.200000,0.212122,0.590435,0.160296,108,1454,1562
4,5,2011-01-05,1,0,1,0,3,1,1,0.226957,0.229270,0.436957,0.186900,82,1518,1600


The dataset contains various columns, including:

- **instant**: Record index.
- **dteday**: Date of the record.
- **season**: The season in which the data was recorded (1: Spring, 2: Summer, 3: Fall, 4: Winter).
- **yr**: Year of the data (0: 2011, 1: 2012).
- **mnth**: Month of the year (1 to 12).
- **holiday**: Whether the day is a holiday (1: Yes, 0: No).
- **weekday**: Day of the week (0: Sunday, 6: Saturday).
- **workingday**: Whether the day is a working day (1: Yes, 0: No).
- **weathersit**: Weather situation (1: Clear, 2: Mist, 3: Light rain/snow, 4: Heavy rain/snow).
- **temp** : Normalized temperature in Celsius. The values are divided to 41 (max).
- **atemp**: Normalized feeling temperature in Celsius. The values are divided to 50 (max).
- **hum**: Normalized humidity. The values are divided to 100 (max).
- **windspeed**: Normalized wind speed. The values are divided to 67 (max).
- **casual**: Count of casual users.
- **registered**: Count of registered users.
- **cnt**: Count of total rental bikes including both casual and registered.

By examining the first few rows, we can get an overview of the dataset before performing further analysis.


#### Load hourly_df

In [3]:
hourly_df = pd.read_csv("https://raw.githubusercontent.com/muchammadwr/bike-sharing-demand-analysis/refs/heads/main/submission/data/hour.csv")
hourly_df.head()

,instant,dteday,season,yr,mnth,hr,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
0,1,2011-01-01,1,0,1,0,0,6,0,1,0.24,0.2879,0.81,0.0,3,13,16
1,2,2011-01-01,1,0,1,1,0,6,0,1,0.22,0.2727,0.80,0.0,8,32,40
2,3,2011-01-01,1,0,1,2,0,6,0,1,0.22,0.2727,0.80,0.0,5,27,32
3,4,2011-01-01,1,0,1,3,0,6,0,1,0.24,0.2879,0.75,0.0,3,10,13
4,5,2011-01-01,1,0,1,4,0,6,0,1,0.24,0.2879,0.75,0.0,0,1,1


This dataset contains similar features to the daily dataset (`day_df`), but with hourly records instead of daily aggregates. Some key columns include:

- **instant**: Unique identifier for each row.
- **dteday**: Date of the record.
- **hr**: Hour of the day (0 to 23).
- **season, yr, mnth, holiday, weekday, workingday, weathersit**: Similar to the daily dataset, representing seasonal and time-related information.
- **temp, atemp, hum, windspeed**: Weather-related features.
- **casual, registered, cnt**: Number of casual users, registered users, and total bike rentals for each hour.

Since this dataset provides a more granular view with hourly records, it allows for a deeper analysis of bike-sharing trends throughout the day.


### Assessing Data

#### Identifying daily_df problems

In [4]:
# Day data information
daily_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 731 entries, 0 to 730
Data columns (total 16 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   instant     731 non-null    int64  
 1   dteday      731 non-null    object 
 2   season      731 non-null    int64  
 3   yr          731 non-null    int64  
 4   mnth        731 non-null    int64  
 5   holiday     731 non-null    int64  
 6   weekday     731 non-null    int64  
 7   workingday  731 non-null    int64  
 8   weathersit  731 non-null    int64  
 9   temp        731 non-null    float64
 10  atemp       731 non-null    float64
 11  hum         731 non-null    float64
 12  windspeed   731 non-null    float64
 13  casual      731 non-null    int64  
 14  registered  731 non-null    int64  
 15  cnt         731 non-null    int64  
dtypes: float64(4), int64(11), object(1)
memory usage: 91.5+ KB


- The dataset contains **731 rows** and **16 columns**.
- All columns have **731 non-null values**, meaning there are **no missing values**.
- The dataset consists of different data types:
  - **Integer (`int64`)**: Used for categorical and count-based features such as `season`, `yr`, `mnth`, `holiday`, `weekday`, `workingday`, `weathersit`, `casual`, `registered`, and `cnt`.
  - **Float (`float64`)**: Used for continuous numerical values like `temp`, `atemp`, `hum`, and `windspeed`.
  - **Object (`object`)**: The `dteday` column, which represents dates in string format.

This information helps in preprocessing and deciding if any columns need transformation (e.g., converting `dteday` to a proper datetime format).


In [5]:
# Check Duplicate the day data
print(daily_df.duplicated().sum())

0


The output is **0**, which means there are no duplicate records in the dataset. This confirms that each row represents a unique entry, and no additional cleaning is required for duplicates.


In [6]:
# Check Missing Value the day data
daily_df.isna().sum()

,0
instant,0
dteday,0
season,0
yr,0
mnth,0
holiday,0
weekday,0
workingday,0
weathersit,0
temp,0


The output shows that **all columns have 0 missing values**, meaning the dataset is complete and does not require imputation or removal of missing data.

Since there are no missing values, we can proceed with the analysis without additional preprocessing for missing data.


In [7]:
# The descriptive statistic from day data
daily_df.describe()

,instant,season,yr,mnth,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
count,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000
mean,366.000000,2.496580,0.500684,6.519836,0.028728,2.997264,0.683995,1.395349,0.495385,0.474354,0.627894,0.190486,848.176471,3656.172367,4504.348837
std,211.165812,1.110807,0.500342,3.451913,0.167155,2.004787,0.465233,0.544894,0.183051,0.162961,0.142429,0.077498,686.622488,1560.256377,1937.211452
min,1.000000,1.000000,0.000000,1.000000,0.000000,0.000000,0.000000,1.000000,0.059130,0.079070,0.000000,0.022392,2.000000,20.000000,22.000000
25%,183.500000,2.000000,0.000000,4.000000,0.000000,1.000000,0.000000,1.000000,0.337083,0.337842,0.520000,0.134950,315.500000,2497.000000,3152.000000
50%,366.000000,3.000000,1.000000,7.000000,0.000000,3.000000,1.000000,1.000000,0.498333,0.486733,0.626667,0.180975,713.000000,3662.000000,4548.000000
75%,548.500000,3.000000,1.000000,10.000000,0.000000,5.000000,1.000000,2.000000,0.655417,0.608602,0.730209,0.233214,1096.000000,4776.500000,5956.000000
max,731.000000,4.000000,1.000000,12.000000,1.000000,6.000000,1.000000,3.000000,0.861667,0.840896,0.972500,0.507463,3410.000000,6946.000000,8714.000000


To understand the overall distribution of numerical variables, we use the `.describe()` function. This provides key statistical summaries, including:

- **count**: Number of non-null values in each column.
- **mean**: Average value of each column.
- **std (standard deviation)**: Measures data dispersion from the mean.
- **min and max**: Minimum and maximum values.
- **25%, 50% (median), and 75%**: Quartiles, showing data distribution.

Key Observations:

- The dataset consists of **731 records**.
- The **average number of total bike rentals (`cnt`) is 4504**, with a **minimum of 22** and a **maximum of 8714**.
- The **temperature (`temp`) ranges from 0.06 to 0.86** (normalized values).
- The dataset is **balanced across years (`yr`)**, with values of `0` (2011) and `1` (2012).
- The **working day (`workingday`) and holiday (`holiday`) columns are binary**, indicating whether a day is a working day or holiday.

This summary helps in understanding the overall trends and potential data distributions before further analysis.


#### Identifying hour_df problems


In [8]:
# Hour data information
hourly_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 17379 entries, 0 to 17378
Data columns (total 17 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   instant     17379 non-null  int64  
 1   dteday      17379 non-null  object 
 2   season      17379 non-null  int64  
 3   yr          17379 non-null  int64  
 4   mnth        17379 non-null  int64  
 5   hr          17379 non-null  int64  
 6   holiday     17379 non-null  int64  
 7   weekday     17379 non-null  int64  
 8   workingday  17379 non-null  int64  
 9   weathersit  17379 non-null  int64  
 10  temp        17379 non-null  float64
 11  atemp       17379 non-null  float64
 12  hum         17379 non-null  float64
 13  windspeed   17379 non-null  float64
 14  casual      17379 non-null  int64  
 15  registered  17379 non-null  int64  
 16  cnt         17379 non-null  int64  
dtypes: float64(4), int64(12), object(1)
memory usage: 2.3+ MB


To understand the structure of the hourly dataset, we use the `.info()` function. This provides an overview of the dataset, including:

- **Total entries**: 17,379 rows.
- **Total columns**: 17 attributes.
- **Data types**:
  - `int64`: 12 columns (e.g., `season`, `yr`, `mnth`, `hr`, `holiday`, etc.).
  - `float64`: 4 columns (`temp`, `atemp`, `hum`, `windspeed`).
  - `object`: 1 column (`dteday`, representing dates).
- **Memory usage**: 2.3+ MB.

Key Observations:

- The dataset contains **hourly records of bike rentals** over two years.
- There are **no missing values** in any column, which simplifies data preprocessing.
- The `hr` column represents hours of the day (0-23), allowing for time-based analysis.
- `cnt` is the total number of rentals, obtained from the sum of `casual` and `registered` users.

This summary helps in understanding the dataset before performing further exploratory data analysis.


In [9]:
# Check duplicate the hour data
print(hourly_df.duplicated().sum())

0


The output is **0**, meaning there are **no duplicate entries** in the dataset. This confirms that each row represents a unique hourly record without redundancy.


In [10]:
# Check Missing Value the hour data
hourly_df.isnull().sum()

,0
instant,0
dteday,0
season,0
yr,0
mnth,0
hr,0
holiday,0
weekday,0
workingday,0
weathersit,0


The output shows **0 missing values** in all 17 columns. This means the dataset is **fully complete**, with no need for imputation or data cleaning related to missing values.


In [11]:
# The descriptive statistic from day data
hourly_df.describe().round(2)

,instant,season,yr,mnth,hr,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
count,17379.00,17379.00,17379.0,17379.00,17379.00,17379.00,17379.00,17379.00,17379.00,17379.00,17379.00,17379.00,17379.00,17379.00,17379.00,17379.00
mean,8690.00,2.50,0.5,6.54,11.55,0.03,3.00,0.68,1.43,0.50,0.48,0.63,0.19,35.68,153.79,189.46
std,5017.03,1.11,0.5,3.44,6.91,0.17,2.01,0.47,0.64,0.19,0.17,0.19,0.12,49.31,151.36,181.39
min,1.00,1.00,0.0,1.00,0.00,0.00,0.00,0.00,1.00,0.02,0.00,0.00,0.00,0.00,0.00,1.00
25%,4345.50,2.00,0.0,4.00,6.00,0.00,1.00,0.00,1.00,0.34,0.33,0.48,0.10,4.00,34.00,40.00
50%,8690.00,3.00,1.0,7.00,12.00,0.00,3.00,1.00,1.00,0.50,0.48,0.63,0.19,17.00,115.00,142.00
75%,13034.50,3.00,1.0,10.00,18.00,0.00,5.00,1.00,2.00,0.66,0.62,0.78,0.25,48.00,220.00,281.00
max,17379.00,4.00,1.0,12.00,23.00,1.00,6.00,1.00,4.00,1.00,1.00,1.00,0.85,367.00,886.00,977.00


To understand the distribution of numerical features, we use the `.describe().round(2)` function. This provides key statistical summaries, including count, mean, standard deviation, min, max, and quartiles.

Key Observations:

- **Total Entries**: 17,379 rows.
- **Hourly Distribution** (`hr`): The values range from **0 to 23**, covering all hours in a day.
- **Temperature (`temp`) & Feels-Like Temperature (`atemp`)**:
  - `temp` (Normalized between 0 and 1) has a **mean of 0.50**, suggesting an even spread over different weather conditions.
  - `atemp` follows a similar distribution with a mean of **0.48**.
- **Humidity (`hum`) & Wind Speed (`windspeed`)**:
  - Humidity ranges from **0 to 1**, with a mean of **0.63**.
  - Wind speed has an average of **0.19**, indicating most days experience mild winds.
- **Bike Rentals (`cnt`)**:
  - The total count (`cnt`) ranges from **1 to 977** per hour.
  - Median rentals per hour is **142**, while the upper quartile reaches **281**, indicating some peak hours with high demand.
  - `casual` riders (mean **35.68**) are significantly fewer than `registered` riders (mean **153.79**), suggesting that most users are registered members.

Conclusion:  
This statistical summary provides insight into the distribution and variability of bike rentals across different hours, weather conditions, and user types. These insights will guide further analysis and model building.


### Cleaning Data

#### Fixing daily_df problem

In [12]:
daily_df.head()

,instant,dteday,season,yr,mnth,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
0,1,2011-01-01,1,0,1,0,6,0,2,0.344167,0.363625,0.805833,0.160446,331,654,985
1,2,2011-01-02,1,0,1,0,0,0,2,0.363478,0.353739,0.696087,0.248539,131,670,801
2,3,2011-01-03,1,0,1,0,1,1,1,0.196364,0.189405,0.437273,0.248309,120,1229,1349
3,4,2011-01-04,1,0,1,0,2,1,1,0.200000,0.212122,0.590435,0.160296,108,1454,1562
4,5,2011-01-05,1,0,1,0,3,1,1,0.226957,0.229270,0.436957,0.186900,82,1518,1600


In [13]:
# Change the column names
daily_columns = {
    "dteday": "date",
    "yr": "year",
    "mnth": "month",
    "temp": "temperature",
    "atemp": "feels_like_temperature",
    "hum": "humidity",
    "cnt": "count",
}

daily_df.rename(columns=daily_columns, inplace=True)

In [14]:
# Fix the datetime data
daily_df["date"] = pd.to_datetime(daily_df["date"])

In [15]:
# Categorical season column
season = {1: "Springer", 2: "Summer", 3: "Fall", 4: "Winter"}

daily_df["season"] = daily_df["season"].replace(season)

In [16]:
# Year column
year = {0: 2011, 1: 2012}

daily_df["year"] = daily_df["year"].replace(year)

In [17]:
# Month column
month = {
    1: "January",
    2: "February",
    3: "March",
    4: "April",
    5: "May",
    6: "June",
    7: "July",
    8: "August",
    9: "September",
    10: "October",
    11: "November",
    12: "December",
}

daily_df["month"] = daily_df["month"].replace(month)

In [18]:
# Change holiday column to categorical values
holiday = {0: "No", 1: "Yes"}

daily_df["holiday"] = daily_df["holiday"].replace(holiday)

In [19]:
# Change coloumns weekdays from numerical to categorical
weekdays = {
    0: "Sunday",
    1: "Monday",
    2: "Tuesday",
    3: "Wednesday",
    4: "Thursday",
    5: "Friday",
    6: "Saturday",
}

daily_df["weekday"] = daily_df["weekday"].replace(weekdays)

In [20]:
# Change coloumns workingday from numerical to categorical
workingday = {0: "No", 1: "Yes"}

daily_df["workingday"] = daily_df["workingday"].replace(workingday)

In [21]:
# Change the weathersit from numerical to categorical
weathersit_labels = {
    1: "Clear",
    2: "Cloudy or Misty",
    3: "Light Rain/Snow",
    4: "Heavy Rain/Snow",
}

daily_df["weathersit"] = daily_df["weathersit"].map(weathersit_labels)

In [22]:
# Denormalized temperature column with multiple by 41
daily_df["temperature"] = daily_df["temperature"] * 41

In [23]:
# Denormalized feels_like_temperature column with multiple by 50
daily_df["feels_like_temperature"] = daily_df["feels_like_temperature"] * 50

In [24]:
# Denormalized windspeed column with multiple by 67
daily_df["windspeed"] = daily_df["windspeed"] * 67

In [25]:
daily_df.head()

,instant,date,season,year,month,holiday,weekday,workingday,weathersit,temperature,feels_like_temperature,humidity,windspeed,casual,registered,count
0,1,2011-01-01,Springer,2011,January,No,Saturday,No,Cloudy or Misty,14.110847,18.18125,0.805833,10.749882,331,654,985
1,2,2011-01-02,Springer,2011,January,No,Sunday,No,Cloudy or Misty,14.902598,17.68695,0.696087,16.652113,131,670,801
2,3,2011-01-03,Springer,2011,January,No,Monday,Yes,Clear,8.050924,9.47025,0.437273,16.636703,120,1229,1349
3,4,2011-01-04,Springer,2011,January,No,Tuesday,Yes,Clear,8.200000,10.60610,0.590435,10.739832,108,1454,1562
4,5,2011-01-05,Springer,2011,January,No,Wednesday,Yes,Clear,9.305237,11.46350,0.436957,12.522300,82,1518,1600


#### Fixing hourly_df problem

In [26]:
hourly_df.head()

,instant,dteday,season,yr,mnth,hr,holiday,weekday,workingday,weathersit,temp,atemp,hum,windspeed,casual,registered,cnt
0,1,2011-01-01,1,0,1,0,0,6,0,1,0.24,0.2879,0.81,0.0,3,13,16
1,2,2011-01-01,1,0,1,1,0,6,0,1,0.22,0.2727,0.80,0.0,8,32,40
2,3,2011-01-01,1,0,1,2,0,6,0,1,0.22,0.2727,0.80,0.0,5,27,32
3,4,2011-01-01,1,0,1,3,0,6,0,1,0.24,0.2879,0.75,0.0,3,10,13
4,5,2011-01-01,1,0,1,4,0,6,0,1,0.24,0.2879,0.75,0.0,0,1,1


In [27]:
# Change the column names
hourly_column = {
    "dteday": "date",
    "yr": "year",
    "mnth": "month",
    "hr": "hour",
    "temp": "temperature",
    "atemp": "feels_like_temperature",
    "hum": "humidity",
    "cnt": "count",
}

hourly_df.rename(columns=hourly_column, inplace=True)
hourly_df.head()

,instant,date,season,year,month,hour,holiday,weekday,workingday,weathersit,temperature,feels_like_temperature,humidity,windspeed,casual,registered,count
0,1,2011-01-01,1,0,1,0,0,6,0,1,0.24,0.2879,0.81,0.0,3,13,16
1,2,2011-01-01,1,0,1,1,0,6,0,1,0.22,0.2727,0.80,0.0,8,32,40
2,3,2011-01-01,1,0,1,2,0,6,0,1,0.22,0.2727,0.80,0.0,5,27,32
3,4,2011-01-01,1,0,1,3,0,6,0,1,0.24,0.2879,0.75,0.0,3,10,13
4,5,2011-01-01,1,0,1,4,0,6,0,1,0.24,0.2879,0.75,0.0,0,1,1


In [28]:
# Fix the datetime data
hourly_df["date"] = pd.to_datetime(hourly_df["date"])

In [29]:
# Categorical season column
season = {1: "Springer", 2: "Summer", 3: "Fall", 4: "Winter"}

hourly_df["season"] = hourly_df["season"].replace(season)

In [30]:
# Year column
year = {0: 2011, 1: 2012}

hourly_df["year"] = hourly_df["year"].replace(year)

In [31]:
# Month column
month = {
    1: "January",
    2: "February",
    3: "March",
    4: "April",
    5: "May",
    6: "June",
    7: "July",
    8: "August",
    9: "September",
    10: "October",
    11: "November",
    12: "December",
}

hourly_df["month"] = hourly_df["month"].replace(month)

In [32]:
# Hour Column
hourly_df["hour"] = hourly_df["hour"].apply(lambda x: time(x, 0))

In [33]:
hourly_df.head()

,instant,date,season,year,month,hour,holiday,weekday,workingday,weathersit,temperature,feels_like_temperature,humidity,windspeed,casual,registered,count
0,1,2011-01-01,Springer,2011,January,00:00:00,0,6,0,1,0.24,0.2879,0.81,0.0,3,13,16
1,2,2011-01-01,Springer,2011,January,01:00:00,0,6,0,1,0.22,0.2727,0.80,0.0,8,32,40
2,3,2011-01-01,Springer,2011,January,02:00:00,0,6,0,1,0.22,0.2727,0.80,0.0,5,27,32
3,4,2011-01-01,Springer,2011,January,03:00:00,0,6,0,1,0.24,0.2879,0.75,0.0,3,10,13
4,5,2011-01-01,Springer,2011,January,04:00:00,0,6,0,1,0.24,0.2879,0.75,0.0,0,1,1


In [34]:
# Change holiday column to categorical values
holiday = {0: "No", 1: "Yes"}

hourly_df["holiday"] = hourly_df["holiday"].replace(holiday)

In [35]:
# Change coloumns weekdays from numerical to categorical
weekdays = {
    0: "Sunday",
    1: "Monday",
    2: "Tuesday",
    3: "Wednesday",
    4: "Thursday",
    5: "Friday",
    6: "Saturday",
}

hourly_df["weekday"] = hourly_df["weekday"].replace(weekdays)

In [36]:
# Change coloumns workingday from numerical to categorical
workingday = {0: "No", 1: "Yes"}

hourly_df["workingday"] = hourly_df["workingday"].replace(workingday)

In [37]:
# Change the weathersit from numerical to categorical
weathersit_labels = {
    1: "Clear",
    2: "Cloudy or Misty",
    3: "Light Rain/Snow",
    4: "Heavy Rain/Snow",
}

hourly_df["weathersit"] = hourly_df["weathersit"].map(weathersit_labels)

In [38]:
# Denormalized temperature column with multiple by 41
hourly_df["temperature"] = hourly_df["temperature"] * 41

In [39]:
# Denormalized feels_like_temperature column with multiple by 50
hourly_df["feels_like_temperature"] = hourly_df["feels_like_temperature"] * 50

In [40]:
# Denormalized humidity column with multiple by 100
hourly_df["humidity"] = hourly_df["humidity"] * 100

In [41]:
# Denormalized windspeed column with multiple by 67
hourly_df["windspeed"] = hourly_df["windspeed"] * 67

In [42]:
hourly_df.head()

,instant,date,season,year,month,hour,holiday,weekday,workingday,weathersit,temperature,feels_like_temperature,humidity,windspeed,casual,registered,count
0,1,2011-01-01,Springer,2011,January,00:00:00,No,Saturday,No,Clear,9.84,14.395,81.0,0.0,3,13,16
1,2,2011-01-01,Springer,2011,January,01:00:00,No,Saturday,No,Clear,9.02,13.635,80.0,0.0,8,32,40
2,3,2011-01-01,Springer,2011,January,02:00:00,No,Saturday,No,Clear,9.02,13.635,80.0,0.0,5,27,32
3,4,2011-01-01,Springer,2011,January,03:00:00,No,Saturday,No,Clear,9.84,14.395,75.0,0.0,3,10,13
4,5,2011-01-01,Springer,2011,January,04:00:00,No,Saturday,No,Clear,9.84,14.395,75.0,0.0,0,1,1


## Exploratory Data Analysis (EDA)

### Explore daily_df


In [43]:
daily_df.describe(include="all")

,instant,date,season,year,month,holiday,weekday,workingday,weathersit,temperature,feels_like_temperature,humidity,windspeed,casual,registered,count
count,731.000000,731,731,731.000000,731,731,731,731,731,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000,731.000000
unique,NaN,NaN,4,NaN,12,2,7,2,3,NaN,NaN,NaN,NaN,NaN,NaN,NaN
top,NaN,NaN,Fall,NaN,January,No,Saturday,Yes,Clear,NaN,NaN,NaN,NaN,NaN,NaN,NaN
freq,NaN,NaN,188,NaN,62,710,105,500,463,NaN,NaN,NaN,NaN,NaN,NaN,NaN
mean,366.000000,2012-01-01 00:00:00,NaN,2011.500684,NaN,NaN,NaN,NaN,NaN,20.310776,23.717699,0.627894,12.762576,848.176471,3656.172367,4504.348837
min,1.000000,2011-01-01 00:00:00,NaN,2011.000000,NaN,NaN,NaN,NaN,NaN,2.424346,3.953480,0.000000,1.500244,2.000000,20.000000,22.000000
25%,183.500000,2011-07-02 12:00:00,NaN,2011.000000,NaN,NaN,NaN,NaN,NaN,13.820424,16.892125,0.520000,9.041650,315.500000,2497.000000,3152.000000
50%,366.000000,2012-01-01 00:00:00,NaN,2012.000000,NaN,NaN,NaN,NaN,NaN,20.431653,24.336650,0.626667,12.125325,713.000000,3662.000000,4548.000000
75%,548.500000,2012-07-01 12:00:00,NaN,2012.000000,NaN,NaN,NaN,NaN,NaN,26.872076,30.430100,0.730209,15.625371,1096.000000,4776.500000,5956.000000
max,731.000000,2012-12-31 00:00:00,NaN,2012.000000,NaN,NaN,NaN,NaN,NaN,35.328347,42.044800,0.972500,34.000021,3410.000000,6946.000000,8714.000000


Based on the descriptive statistics:

- The dataset contains **731 daily records** covering the period from **January 1, 2011, to December 31, 2012**, with no missing values.
- The average daily bike rentals were **4,504**, consisting of approximately **848 casual users** and **3,656 registered users**.
- Daily rental demand ranged from **22** to **8,714** rentals, indicating substantial variation in usage.
- The average temperature was **20.31°C**, with humidity averaging **62.8%** and wind speed averaging **12.76 km/h**.
- **Fall** was the most frequent season, **January** appeared most often among months, and **Clear** weather was the most common weather condition.


In [44]:
# numerical column values
num = daily_df.select_dtypes(
    exclude=["object", "category", "datetime"]
).columns.tolist()
daily_df[num].head()

,instant,year,temperature,feels_like_temperature,humidity,windspeed,casual,registered,count
0,1,2011,14.110847,18.18125,0.805833,10.749882,331,654,985
1,2,2011,14.902598,17.68695,0.696087,16.652113,131,670,801
2,3,2011,8.050924,9.47025,0.437273,16.636703,120,1229,1349
3,4,2011,8.200000,10.60610,0.590435,10.739832,108,1454,1562
4,5,2011,9.305237,11.46350,0.436957,12.522300,82,1518,1600


In [45]:
# datetype column values
num = daily_df.select_dtypes(include=["datetime"]).columns.tolist()
daily_df[num].head()

,date
0,2011-01-01
1,2011-01-02
2,2011-01-03
3,2011-01-04
4,2011-01-05


In [47]:
# categorical column values
num = daily_df.select_dtypes(
    include=["object", "category"]
).columns.tolist()
daily_df[num].head()

,season,month,holiday,weekday,workingday,weathersit
0,Springer,January,No,Saturday,No,Cloudy or Misty
1,Springer,January,No,Sunday,No,Cloudy or Misty
2,Springer,January,No,Monday,Yes,Clear
3,Springer,January,No,Tuesday,Yes,Clear
4,Springer,January,No,Wednesday,Yes,Clear


In [106]:
# Rented by season
daily_df.groupby("season")["count"].agg({"sum", "mean", "count"}).round(2)

,count,sum,mean
season,,,
Fall,188,1061129,5644.30
Springer,181,471348,2604.13
Summer,184,918589,4992.33
Winter,178,841613,4728.16


Based on the seasonal summary:

- **Fall** recorded the highest average daily rentals (**5,644**), indicating the strongest demand among all seasons.
- **Spring** had the lowest average daily rentals (**2,604**), suggesting lower bike usage during this season.
- **Summer** and **Winter** showed relatively similar rental demand, averaging **4,992** and **4,728** daily rentals, respectively.
- Rental variability (standard deviation) was highest in **Winter** and **Summer**, indicating more fluctuation in daily demand, while **Fall** exhibited the most consistent rental patterns.


In [49]:
# Rented by month
month_order = [
    "January", "February", "March", "April",
    "May", "June", "July", "August",
    "September", "October", "November", "December"
]

daily_df["month"] = pd.Categorical(daily_df["month"], categories=month_order, ordered=True)
daily_df.groupby("month")["count"].agg({"sum", "mean", "count"}).round(2)

,mean,sum,count
month,,,
January,2176.34,134933,62
February,2655.30,151352,57
March,3692.26,228920,62
April,4484.90,269094,60
May,5349.77,331686,62
June,5772.37,346342,60
July,5563.68,344948,62
August,5664.42,351194,62
September,5766.52,345991,60


In [50]:
#  Rented by Holiday
daily_df.groupby("holiday")["count"].agg({"sum", "mean", "count"}).round(2)

,mean,sum,count
holiday,,,
No,4527.1,3214244,710
Yes,3735.0,78435,21


In [51]:
# Rented by Weekday
day_order = [
    "Sunday", "Monday", "Tuesday", "Wednesday",
    "Thursday", "Friday", "Saturday"
]

daily_df["weekday"] = pd.Categorical(daily_df["weekday"], categories=day_order, ordered=True)
daily_df.groupby("weekday")["count"].agg({"sum", "mean", "count"}).round(2)

,mean,sum,count
weekday,,,
Sunday,4228.83,444027,105
Monday,4338.12,455503,105
Tuesday,4510.66,469109,104
Wednesday,4548.54,473048,104
Thursday,4667.26,485395,104
Friday,4690.29,487790,104
Saturday,4550.54,477807,105


Based on the weekday summary:

- **Friday** recorded the highest average daily rentals (**4,690**), followed closely by **Thursday** (**4,667**).
- **Sunday** had the lowest average daily rentals (**4,229**), indicating lower bike usage on weekends.
- Total rentals were relatively consistent across weekdays, ranging from **444,027** (Sunday) to **487,790** (Friday).
- Overall, bike rental demand remained fairly stable throughout the week, with slightly higher activity on weekdays than weekends.


In [164]:
# rented by month
daily_df.groupby("month")["count"].agg({"sum", "mean", "count"}).round(2)

,count,mean,sum
month,,,
April,60,4484.90,269094
August,62,5664.42,351194
December,62,3403.81,211036
February,57,2655.30,151352
January,62,2176.34,134933
July,62,5563.68,344948
June,60,5772.37,346342
March,62,3692.26,228920
May,62,5349.77,331686


Based on the monthly summary:

- **June** recorded the highest average daily rentals (**5,772**), followed closely by **September** (**5,766**) and **August** (**5,664**).
- **January** had the lowest average daily rentals (**2,176**), with **February** also showing relatively low demand (**2,655**).
- Bike rental demand generally increased from **spring to summer**, peaked during the warmer months, and gradually declined toward the end of the year.
- These results suggest that **seasonal and weather conditions** play an important role in influencing bike rental demand.


In [165]:
# Sales performance by years
daily_df.groupby("year")["count"].agg({"sum", "mean", "count"}).round(2)

,count,mean,sum
year,,,
2011,365,3405.76,1243103
2012,366,5599.93,2049576


- **2012** recorded significantly higher bike rental demand than **2011**, with an average of **5,600** daily rentals compared to **3,406** in 2011.
- Total rentals increased from **1.24 million** in **2011** to **2.05 million** in **2012**, representing substantial growth in bike-sharing usage.
- These results indicate a strong upward trend in demand, suggesting increased adoption and utilization of the bike-sharing service over the two-year period.


In [166]:
# registered customers compare to casual customers
daily_df.groupby("weekday").agg(
    {"casual": ["mean", "sum"], "registered": ["mean", "sum"]}
).round(2)

casual         registered        
              mean     sum       mean     sum
weekday                                      
Friday      752.29   78238    3938.00  409552
Monday      674.13   70784    3663.99  384719
Saturday   1465.26  153852    3085.29  323955
Sunday     1338.30  140521    2890.53  303506
Thursday    590.96   61460    4076.30  423935
Tuesday     556.18   57843    3954.48  411266
Wednesday   551.14   57319    3997.39  415729

Based on the weekday comparison:

- **Registered users** consistently contributed more rentals than **casual users** on every day of the week, indicating they are the primary customer segment.
- **Casual users** were most active on **Saturday** (average **1,465** rentals) and **Sunday** (average **1,338** rentals), suggesting higher recreational usage during weekends.
- **Registered users** recorded their highest average rentals on **Thursday** (**4,076**) and **Friday** (**3,938**), reflecting strong commuting demand on weekdays.
- These patterns indicate that **registered users primarily use the service for weekday commuting**, while **casual users are more likely to rent bikes for leisure on weekends**.


### Explore hourly_df


In [167]:
hourly_df.describe(include="all")

,instant,date,season,year,month,hour,holiday,weekday,workingday,weathersit,temperature,feels_like_temperature,humidity,windspeed,casual,registered,count
count,17379.0000,17379,17379,17379.000000,17379,17379,17379,17379,17379,17379,17379.000000,17379.000000,17379.000000,17379.000000,17379.000000,17379.000000,17379.000000
unique,NaN,NaN,4,NaN,12,24,2,7,2,4,NaN,NaN,NaN,NaN,NaN,NaN,NaN
top,NaN,NaN,Fall,NaN,July,16:00:00,No,Saturday,Yes,Clear,NaN,NaN,NaN,NaN,NaN,NaN,NaN
freq,NaN,NaN,4496,NaN,1488,730,16879,2512,11865,11413,NaN,NaN,NaN,NaN,NaN,NaN,NaN
mean,8690.0000,2012-01-02 04:08:34.552045568,NaN,2011.502561,NaN,NaN,NaN,NaN,NaN,NaN,20.376474,23.788755,62.722884,12.736540,35.676218,153.786869,189.463088
min,1.0000,2011-01-01 00:00:00,NaN,2011.000000,NaN,NaN,NaN,NaN,NaN,NaN,0.820000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000
25%,4345.5000,2011-07-04 00:00:00,NaN,2011.000000,NaN,NaN,NaN,NaN,NaN,NaN,13.940000,16.665000,48.000000,7.001500,4.000000,34.000000,40.000000
50%,8690.0000,2012-01-02 00:00:00,NaN,2012.000000,NaN,NaN,NaN,NaN,NaN,NaN,20.500000,24.240000,63.000000,12.998000,17.000000,115.000000,142.000000
75%,13034.5000,2012-07-02 00:00:00,NaN,2012.000000,NaN,NaN,NaN,NaN,NaN,NaN,27.060000,31.060000,78.000000,16.997900,48.000000,220.000000,281.000000
max,17379.0000,2012-12-31 00:00:00,NaN,2012.000000,NaN,NaN,NaN,NaN,NaN,NaN,41.000000,50.000000,100.000000,56.996900,367.000000,886.000000,977.000000


Based on the descriptive statistics:

- The dataset contains **17,379 hourly records** covering the period from **January 1, 2011, to December 31, 2012**, with no missing values.
- The average hourly bike rentals were **189**, consisting of approximately **36 casual users** and **154 registered users**.
- Hourly rental demand ranged from **1** to **977** rentals, indicating substantial variation throughout the day.
- The average temperature was **20.38°C**, with humidity averaging **62.7%** and wind speed averaging **12.74 km/h**.
- **Fall** was the most frequent season, **May** appeared most often among months, **4:00 PM (16:00)** was the most common recorded hour, and **Clear** weather was the predominant weather condition.


In [168]:
# numerical column values
num = hourly_df.select_dtypes(
    exclude=["object", "category", "datetime"]
).columns.tolist()
hourly_df[num].head()

,instant,year,temperature,feels_like_temperature,humidity,windspeed,casual,registered,count
0,1,2011,9.84,14.395,81.0,0.0,3,13,16
1,2,2011,9.02,13.635,80.0,0.0,8,32,40
2,3,2011,9.02,13.635,80.0,0.0,5,27,32
3,4,2011,9.84,14.395,75.0,0.0,3,10,13
4,5,2011,9.84,14.395,75.0,0.0,0,1,1


In [169]:
# datetyme column values
num = hourly_df.select_dtypes(include=["datetime"]).columns.tolist()
hourly_df[num].head()

,date
0,2011-01-01
1,2011-01-01
2,2011-01-01
3,2011-01-01
4,2011-01-01


In [170]:
# Rented by season
hourly_df.groupby("season")["count"].agg({"sum", "mean", "count"}).round(2)

,count,mean,sum
season,,,
Fall,4496,236.02,1061129
Springer,4242,111.11,471348
Summer,4409,208.34,918589
Winter,4232,198.87,841613


Based on the seasonal summary:

* **Fall** recorded the highest average rentals (**236.02**) and total rentals (**1,061,129**), indicating the strongest demand among all seasons.
* **Spring** had the lowest average rentals (**111.11**) and total rentals (**471,348**), suggesting significantly lower bike usage.
* **Summer** ranked second with an average of **208.34 rentals**, followed closely by **Winter** with **198.87 rentals**.

Overall, rental demand increased sharply from **Spring to Fall**, before declining slightly during **Winter**.



In [171]:
# rented by hour
hourly_df.groupby("hour")["count"].agg({"sum", "mean", "count"}).round(2)

,count,mean,sum
hour,,,
00:00:00,726,53.90,39130
01:00:00,724,33.38,24164
02:00:00,715,22.87,16352
03:00:00,697,11.73,8174
04:00:00,697,6.35,4428
05:00:00,717,19.89,14261
06:00:00,725,76.04,55132
07:00:00,727,212.06,154171
08:00:00,727,359.01,261001


Based on the hourly summary:

* Rental demand was lowest during **00:00–05:00**, reaching its minimum at **04:00** with an average of only **6.35 rentals**.
* Demand increased sharply during the morning commuting period, reaching **359.01 average rentals at 08:00**.
* The highest demand occurred at **17:00**, averaging **461.45 rentals** and totaling **336,860 rentals**.
* Demand remained high at **18:00** with an average of **425.51 rentals**, before gradually declining throughout the evening.

Overall, bike rental demand shows clear commuting patterns, with major peaks during the **morning and evening rush hours**.


In [172]:
hourly_df.head()

,instant,date,season,year,month,hour,holiday,weekday,workingday,weathersit,temperature,feels_like_temperature,humidity,windspeed,casual,registered,count
0,1,2011-01-01,Springer,2011,January,00:00:00,No,Saturday,No,Clear,9.84,14.395,81.0,0.0,3,13,16
1,2,2011-01-01,Springer,2011,January,01:00:00,No,Saturday,No,Clear,9.02,13.635,80.0,0.0,8,32,40
2,3,2011-01-01,Springer,2011,January,02:00:00,No,Saturday,No,Clear,9.02,13.635,80.0,0.0,5,27,32
3,4,2011-01-01,Springer,2011,January,03:00:00,No,Saturday,No,Clear,9.84,14.395,75.0,0.0,3,10,13
4,5,2011-01-01,Springer,2011,January,04:00:00,No,Saturday,No,Clear,9.84,14.395,75.0,0.0,0,1,1


In [64]:
# registered customers compare to casual customers
hourly_df.groupby("hour").agg(
    {"casual": ["mean", "sum"], "registered": ["mean", "sum"]}
).round(2)

casual        registered        
           mean    sum       mean     sum
hour                                     
00:00:00  10.16   7375      43.74   31755
01:00:00   6.50   4709      26.87   19455
02:00:00   4.77   3412      18.10   12940
03:00:00   2.72   1893       9.01    6281
04:00:00   1.25    874       5.10    3554
05:00:00   1.41   1012      18.48   13249
06:00:00   4.16   3017      71.88   52115
07:00:00  11.06   8037     201.01  146134
08:00:00  21.68  15761     337.33  245240
09:00:00  30.89  22458     188.42  136980
10:00:00  46.48  33789     127.19   92468
11:00:00  59.54  43286     148.60  108034
12:00:00  68.29  49718     185.02  134696
13:00:00  72.31  52713     181.35  132206
14:00:00  75.57  55089     165.38  120563
15:00:00  74.91  54606     176.33  128543
16:00:00  73.75  53834     238.24  173914
17:00:00  74.27  54220     387.18  282640
18:00:00  61.12  44496     364.39  265276
19:00:00  48.77  35505     262.75  191284
20:00:00  36.23  26378     189.80  138172
21:00:00  28.26  20570     144.06  104875
22:00:00  22.25  16200     109.08   79412
23:00:00  15.20  11065      72.63   52876

## Visualization & Explanatory Analysis

### **Question 1:** During the 2012 period, which hours, days, and seasons recorded the highest average bike rentals?


### **Question 2:** How did monthly rental demand change throughout the 2012 period?



**Insight:** (Opsional)
- xxx
- xxx


### **Question 3:** How did average bike rentals differ across weather conditions during 2012?


### **Question 4:** Which weather-related factor (temperature, humidity, or wind speed) had the strongest relationship with rental demand?


### **Question 5:** How did rental patterns differ between casual and registered users during 2012?


### **Question 6:** Which user segment contributed the largest share of total rentals over the two-year period?

## Analisis Lanjutan (Opsional)

**Insight:** (Opsional)
- xxx
- xxx

## Conclusion & Recommendation

- **Conclusion pertanyaan 1:** xxx
- **Conclusion pertanyaan 2:** xxx
- ...

**Rekomendasi Action Item:**
- xxx
- xxx